### ISSUE TITLE
can't find modules after switch from v0.9.0.4 to v0.9.0.5

### ISSUE THREAD

**Issue by polymorphicengine** [haskell-hint/hint#142]
Hey,
I made some changes to my gui project for tidal recently and decided to publish a new release. I didn't change anything relating to hint, so I thought that the workflow I had should be still fine. I also decicded to use the newest version of hint (previously I used 0.9.0.4).

But I noticed that in the new release hint wasn't able to find any of the haskell modules. This makes me think something changed in the way the interpreter is called with arguments? The arguments I use with my interpreter are:

```
args:: String -> [String]
args lib = ["-clear-package-db", "-package-db", lib ++ "haskell-libs/package.conf.d", "-package-db", lib ++ "haskell-libs/package.db", "-v"]
```

where `lib` is dependent on the location of the binary file. This works fine with hint 0.9.0.4, but is apprently broken with the latest version.

**Comment by gelisam** [comment]
I confirm that I can reproduce this locally.

The only two changes which happened in 0.9.0.5 are to add support for ghc-9.2.1 and to refactor the code to make this support easier to add. Perhaps that refactoring accidentally broke something?

In any case, the first step is to write a test case so that this regression doesn't occur again!

**Comment by gelisam** [comment]
`git bisect` blames a73aceb6413fd57bcdcf9c220f527ea94cb0f920, which is indeed a refactoring which shouldn't have changed anything; especially since it only affects logging!

Pinging @brandon-leapyear, who authored the refactoring. Any idea how that refactoring could possibly affect which package databases the ghc library can see?

**Comment by brandon-leapyear** [comment]
The immediate thing that sticks out to me is I'm not sure why I put the
modifyLogger call at the top instead of right before the `setDynFlags df2`
line. If you move the modifyLogger line down, does that change anything?

**Comment by gelisam** [comment]
That fixes it, thanks!

**Comment by gelisam** [comment]
I have a hypothesis as to how configuring a logger can possibly have an effect on package databases. Here's the relevant code:

    GHC.modifyLogger (const logger)
    ...
    (df2, ...) <- parseDynamicFlags ... ["-package-db=..."]
    GHC.setSessionDynFlags df2

 I'm using ghc-8.4.4, and so `modifyLogger` is implemented in terms of `setSessionDynFlags`:

    df <- getSessionDynFlags
    setSessionDynFlags df{log_action = const logger $ DynFlags.log_action df}
    ...
    (df2, ...) <- runGhc $ parseDynamicFlags ... ["-package-db=..."]
    GHC.setSessionDynFlags df2

Observe that `setSessionDynFlags` is called twice: once without the `-package-db` arguments, and once with them. According to [its documentation](https://hackage.haskell.org/package/ghc-8.8.3/docs/GHC.html#v:setSessionDynFlags), `setSessionDynFlags` does many things, including reading the package database _unless it has already been read_. So I think the first call reads the _default_ package database and sets some `packageDbHasBeenRead` flag indicating that the package databases have been read, then the second call, which should normally be reading the package database specified by the `-package-db` argument, skips reading the package databases because of the `packageDbHasBeenRead` flag.

**Comment by gelisam** [comment]
Now fixed in HEAD; the next step is to release a new version on hackage. And maybe to deprecate the version which has the bug?

**Comment by gelisam** [comment]
0.9.0.6 is now published, and 0.9.0.5 is now deprecated. Thank you both!
